### PR TITLE
Add support for nfp based packing.

### PR DESCRIPTION
**Pull request by pentacular**
This also removes Page and page and obsoletes the 'layout' geometry type.

Much of the complexity comes from supporting merging of geometry coming out of c++ as well as allowing c++ to add tags.

### CODE REVIEW

File: algorithm/cgal/Pack.h [hunk 1]
```diff
@@ -1,49 +1,468 @@
...
+      }
+      auto& pwh = part.pwhs[0];
+      if (pwh.outer_boundary().size() < 1) {
+        return;
+      }
+      part.area = 0;
+      part.pwhs_offset = EK::Point_2(0, 0) - pwh.outer_boundary()[0];
+      move_pwhs(part.pwhs, part.pwhs_offset);
+
+      // The npwhs is a reflection of the pwhs.
```
Comment (BarbourSmith) [inline algorithm/cgal/Pack.h:154]: These look like typos, I know they aren't but they look like it 😂

File: algorithm/cgal/Pack.h [hunk 2]
```diff
@@ -1,49 +1,468 @@
...
+          sheets.push_back(std::move(sheet));
+          std::cout << "Setting pack:sheet on nth=" << nth << std::endl;
+          geometry->setType(nth, GEOMETRY_EMPTY);
           break;
         }
       }
     }
 
+    {
+      // Add an unbounded sheet at the end.
```
Comment (BarbourSmith) [inline algorithm/cgal/Pack.h:273]: Cool idea 👍 👍

File: algorithm/cgal/cast_util.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    const EK::Plane_3& reference_plane, const EK::Vector_3& reference_vector,
+    std::vector<CGAL::Polygon_with_holes_2<EK>>& simple_pwhs) {
+  General_polygon_set_2 gps;
+  std::vector<CGAL::Polygon_with_holes_2<EK>> pwhs;
+  cast_mesh_to_gps(mesh, reference_plane, reference_vector, gps);
+  copy_gps_to_pwhs(gps, pwhs);
+  simplifyPolygonsWithHoles(pwhs, simple_pwhs);
+}
+
+void cast_polygons_with_holes_to_polygons_with_holes(
```
Comment (BarbourSmith) [inline algorithm/cgal/cast_util.h:38]: Why does this need to do anything?

File: algorithm/cgal/surface_mesh_util.h [hunk 4]
```diff
@@ -238,13 +238,17 @@ static void PlanarSurfaceMeshFacetsToPolygonSet(
...
+      auto target = plane.to_2d(mesh.point(mesh.target(edge)));
+      if (source != target) {
+        Segment_2 segment{source, target};
+        insert(arrangement, segment);
+      }
       edge = mesh.next(edge);
     } while (edge != start);
     // The arrangement shouldn't produce polygons with holes, so this might be
     // simplified.
+    // FIX: This is probably completely wrong.
```
Comment (BarbourSmith) [inline algorithm/cgal/surface_mesh_util.h:251]: 😛

**Review by BarbourSmith — APPROVED** [review]
The regression testing images do a really good job here of showing the improvements. Awesome!

File: algorithm/cgal/Pack.h (hunk 1 above)
Comment (pentacular) [reply algorithm/cgal/Pack.h:154]: Yeah, probably should have picked better names. :)

File: algorithm/cgal/cast_util.h (hunk 3 above)
Comment (pentacular) [reply algorithm/cgal/cast_util.h:38]: Ah, they end up projected onto a different plane, potentially.

Comment (BarbourSmith) [reply algorithm/cgal/cast_util.h:38]: That makes sense :)

**Pull request merged**